sql/parser/eval.go

```diff
@@ -1179,35 +1181,39 @@ func ParseDate(s DString) (DDate, error) {
 }
 
 // ParseTimestamp parses the timestamp.
-func ParseTimestamp(s DString) (DTimestamp, error) {
+func (ctx EvalContext) ParseTimestamp(s DString) (DTimestamp, error) {
 	str := string(s)
 	t, err := time.Parse(dateFormat, str)
 	if err == nil {
 		return DTimestamp{Time: t}, nil
 	}
-	t, err = time.Parse(timestampFormat, str)
+	loc := time.UTC
+	if ctx.Offset != 0 {
+		loc = time.FixedZone("", int(ctx.Offset/int64(time.Second)))
+	}
+	t, err = time.ParseInLocation(timestampFormat, str, loc)
 	if err == nil {
 		t = t.UTC()
 		return DTimestamp{Time: t}, nil
 	}
-	t, err = time.Parse(TimestampWithOffsetZoneFormat, str)
+	t, err = time.ParseInLocation(TimestampWithOffsetZoneFormat, str, loc)
 	if err == nil {
 		t = t.UTC()
 		return DTimestamp{Time: t}, nil
 	}
-	t, err = time.Parse(timestampWithNamedZoneFormat, str)
-	if err == nil {
-		// Parsing using a named time zone is imperfect for two reasons:
-		// 1. Some named time zones are ambiguous (PST can be US PST and
-		// phillipines PST), and 2. The code needs to have access to the entire
-		// database of named timed zones in order to get some time offset,
-		// and it's not clear what are the memory requirements for that.
-		// TODO(vivek): Implement SET TIME ZONE to set a time zone and use
-		// time.ParseInLocation()
-		return DummyTimestamp, util.Errorf("TODO(vivek): named time zone input not supported")
+	if ctx.Location != "" {
+		loc, err := time.LoadLocation(ctx.Location)
+		if err != nil {
+			return DummyTimestamp, err
+		}
+		t, err = time.ParseInLocation(timestampWithNamedZoneFormat, str, loc)
```

Comment:
We also need to attempt `timestampFormat` if the time zone was specified as a location name instead of an offset. Reorder this so that we load the location from offset or location if either is given, and then we do all the parsing in that zone. (we could try the `OffsetZoneFormat` first, since that format is independent of the context time zone).